internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1120 @@ package controller
+	csiDriver.Name = r.driver.Name
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&csiDriver), &csiDriver); client.IgnoreNotFound(err) != nil {
+		r.log.Error(err, "Failed to query the existence of a CSI Driver")
+		return err
+	}
+	if csiDriver.UID != "" {
+		ownerObjKey := client.ObjectKey{}
+		if ownerRef := csiDriver.GetAnnotations()[ownerRefAnnotationKey]; ownerRef != "" {
+			if err := json.Unmarshal([]byte(ownerRef), &ownerObjKey); err != nil {
+				r.log.Error(err, "Failed to parse owner ref annotation on CSI Driver")
+				return err
+			}
+		}
+		if csiDriver.Namespace != r.driver.Namespace || csiDriver.Name != r.driver.Name {
+			err := fmt.Errorf("invalid driver name")
+			r.log.Error(err, "Desired name already in use by a different CSI Driver")
+		}
+	}
+
+	// Load operator configuration resource
+	opConfig := csiv1a1.OperatorConfig{}
+	opConfig.Name = operatorConfigName
+	opConfig.Namespace = operatorNamespace
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&opConfig), &opConfig); client.IgnoreNotFound(err) != nil {
+		r.log.Error(err, "Unable to load operatorconfig.csi.ceph.io", "name", client.ObjectKeyFromObject(&opConfig))
+		return err
+	}
+
+	// Extract the driver sort name and driver type
+	matches := nameRegExp.FindStringSubmatch(r.driver.Name)
+	if len(matches) != 3 {
+		return fmt.Errorf("invalid driver name")
+	}
+	r.driverPrefix = matches[1]
+	r.driverType = DriverType(strings.ToLower(matches[2]))
+
+	// Load the current desired state in the form of a ceph csi driver resource
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&r.driver), &r.driver); err != nil {
+		r.log.Error(err, "Unable to load driver.csi.ceph.io object", "name", client.ObjectKeyFromObject(&r.driver))
+		return client.IgnoreNotFound(err)
+	}
+
+	// Creating a copy of the driver spec, making sure any local changes will not effect the object residing
+	// in the client's cache
+	r.driver.Spec = *r.driver.Spec.DeepCopy()
+	mergeDriverSpecs(&r.driver.Spec, opConfig.Spec.DriverSpecDefaults)
+
+	r.images = maps.Clone(imageDefaults)
+	if r.driver.Spec.ImageSet != nil {
+		imageSetCM := corev1.ConfigMap{}
+		imageSetCM.Name = r.driver.Spec.ImageSet.Name
+		imageSetCM.Namespace = r.driver.Namespace
+		if err := r.Get(r.ctx, client.ObjectKeyFromObject(&imageSetCM), &imageSetCM); err != nil {
+			r.log.Error(err, "Unable to load driver specified image set config map", "name", client.ObjectKeyFromObject(&imageSetCM))
+			return err
+		}
+
+		maps.Copy(r.images, imageSetCM.Data)
+	}
+
+	return nil
+}
+
+func (r *driverReconcile) reconcileK8sCsiDriver() error {
+	existingCsiDriver := &storagev1.CSIDriver{}
+	existingCsiDriver.Name = r.driver.Name
+
+	log := r.log.WithValues("driverName", existingCsiDriver.Name)
+	log.Info("Reconciling CSI Driver")
+
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(existingCsiDriver), existingCsiDriver); client.IgnoreNotFound(err) != nil {
+		log.Error(err, "Failed to load CSI Driver resource")
+		return err
```

Comment:
general observation, I see logging and returning error a lot, I believe logging is handling error and the current pattern duplicates the message at func level and at reconcile level.

if we still want to add context, maybe only `return fmt.Errorf("our reason", err)`?

Reply:
The pattern you are suggesting can be interpreted in two different ways - both of the interpretations unfavorable):
1. All errors are just generic errors with a different message - This is a very bad result as it makes writing tests and log analyzers work much harder.
2. Each error message defines a new unique error type - This interpretation is basically analogous to #1 and has the exact same set of problem 

The favorite option is none of the above, it is creating (or reusing) a closed set of Error types that have meaningful fields to communicate the needed information. For reconcilers the set of k8serror is good enough as it already captures  90% of what we need to communicate. For the remaining 10% I will introduce our own error types in a future PR

Regarding the logging aspects: I agree that we need to relook at our strategy around error logging, But that is for another time and another PR

Reply:
resolve.